opacus/grad_sample/grad_sample_module.py

```diff
@@ -100,6 +102,42 @@ def add_hooks(self, loss_reduction: str = "mean", batch_first: bool = True) -> N
                 )
         self.enable_hooks()
 
+    def add_ddp_hook(
+        self, engine, loss_reduction: str = "mean", batch_first: bool = True
+    ) -> None:
+        """
+        Special function to enable DDP support on top of a GradSampleModule
+        """
+
+        # self.n_params = len(
+        #     [n for n, p in self.module.named_parameters() if p.requires_grad]
+        # )
+
+        # We store the number of layers for the per-layer clipping
+        self.n_params = 0
+        self.ddp_hook_activated = True
+
+        # TODO: should we use self._module.parameters() instead?
+
+        # for module in self.trainable_modules():
+        #     if type(module) in self.GRAD_SAMPLERS:
+        #         self.n_params += 1
+        #         self.autograd_grad_sample_hooks.append(
+        #             module.register_hook(
+        #                 partial(
+        #                     self.ddp_backward_callback,
+        #                     engine,
+        #                     module,
+        #                 )
+        #             )
+        #         )
+
+        # We iterate over the parameters and not the submodules
+        for p in self.parameters():
+            if p.requires_grad:
+                self.n_params += 1
+                p.register_hook(partial(self.ddp_backward_callback, engine, p))
+
     def remove_hooks(self) -> None:
```

Comment:
Do we need to update `remove_hooks` to account for new hooks for the proper cleanup?

Reply:
Indeed, I'll add a helper once we clarify the architecture